### ISSUE TITLE
Flickering in highly lit areas on Mac in FFXIV

### ISSUE THREAD

**Issue by tyrexis26** [KhronosGroup/MoltenVK#1542]
Flickering is an issue in areas of high light, such as Mist and Limsa Lominsa in FFXIV.  Adjustments/support may be needed for occlusion culling.

**Comment by marzent** [comment]
Seems to be fine with 1.1.1 and lower and happens with 1.1.2 and onwards...

It's probably related to 55d206e058bb17f2c6fab718c939119e1f1a30f3 and 8e8edbadb10672c1d175f52b422536b33b69ca2d since changing occlusion settings in-game makes the flickering actually go away for a few seconds after which it gradually returns

**Comment by marzent** [comment]
If it would help I could also quickly record a video of it happening

**Comment by cdavis5e** [comment]
The irony is that I initially wrote those patches trying to *fix* occlusion problems in FFXIV... /me sighs

**Comment by tyrexis26** [comment]
Your work is very much appreciated - without it, I doubt I would be playing on my Mac.  Sometimes the flickering gives me a headache, though, and I'm not really sure what to do about it.

**Comment by tyrexis26** [comment]
(sorry about closing it, hit the wrong button)

**Comment by FleetAdmiralButter** [comment]
Just backing up @marzent's point about occlusion:

When standing in the Crystal Tower, the flickering is especially bad when Occlusion Culling is enabled. Disabling Culling does improve it in this area:

https://user-images.githubusercontent.com/37036084/158099752-2b77d4c4-d9f1-454a-9431-5acd455743a8.mp4

However, even with Culling disabled, the flickering will still occur in those areas depending on light source:

https://user-images.githubusercontent.com/37036084/158099813-745c6e0c-f86b-415a-b593-4ca3eceee75a.mp4

**Comment by marzent** [comment]
For me on a 6900XT I actually barely notice any changes between turning Occlusion Culling on and off, however the flickering is also faster (looks like it is tied to framerate though)

**Comment by seathasky** [comment]
Can confirm, flickering happens on bot my RX 570 Mac system and 5700XT Mac in FF14 in all versions past 1.1.2 The only version it doesnt, is CX's mvk 1.1.0, has no source code.

**Comment by Ceolona** [comment]
Adding a “me too” to the list. M1 MBA.  My experience seems to have it affected by Occlusion Culling, Glare, and camera angle.

**Comment by jakeaw** [comment]
I'm having the same issue.  M1 Max

**Comment by seathasky** [comment]
@billhollings Any word on this?

**Comment by seathasky** [comment]
So I guess this is not going to be fixed?

**Comment by seathasky** [comment]
@billhollings Can we get an update on this? It's very important.

**Comment by billhollings** [comment]
> @billhollings Can we get an update on this? It's very important.

I'll need to get set up to evaluate this. Can someone point me to how to install the FFXIV environment you're all running, on an M1 machine, please?

I'm also curious about why this is appearing now, since the MoltenVK updates that are being mentioned above are over a year old.

**Comment by zebe** [comment]
To answer your second question first: The official Square Enix FFXIV client for macOS is getting painfully outdated to the point of not working at all in certain configurations, so more people have been turning to solutions like Crossover and, specifically, XIV on Mac. Those use the newer updates of MoltenVK, whereas the official client is still using a very old build. Between the recent issues with the official client (it was completely unable to launch on macOS 12.3.x for a while), and the advent of XIV on Mac, there is a sudden influx of Mac users moving from very old MoltenVK to a modern build.

As for how to get set up to see this yourself:
1. I would download the FFXIV free trial and make an account: https://na.finalfantasyxiv.com/mac/download/
2. Again, this client will have a very old MoltenVK so to reproduce the issue you'd need to then set up XIV on Mac: https://www.xivmac.com
3. XIV on Mac has a checkbox in setup for "Modern MoltenVK" so you would need to enable that to see this problem. It defaults to using the very old version similar to the official client, partly because of this bug.

Probably anyone in this thread would be willing to provide more detailed setup assistance or a screen recording/logs of the issue upon request, as well.

**Comment by seathasky** [comment]
XIV on Mac dev here. That's correct. We're using an older version of MVK as default in our launcher. Jeremy at Codweavers gave us permission to distribute this. It's a custom compile from a really old version of CX21. It fixes flickering and also fixes stuttering on RX 500 GPUs, but is not without issue. Crashing in certain game areas is one of it's faults. 

It is also why we have the option for a modern MVK in our launcher (even though modern causes FF14 to flicker, modern fixes crashing in certain areas that the CX compiles does not.

Basically, we'd love to just have one mvk to use, an all around compatible MVK that fixes both flickering, rx 500 series issues and crashing.

**Comment by cdavis5e** [comment]
> It's very unfortunate situation for Mac FF14 players right now. Square Enix native mac launcher does not work on Monterey nor has it been updated in 2 1/2 years.

I had a view from inside CW before getting laid off, and... it's probably my fault.

We tried updating MoltenVK to a newer version, but we encountered all sorts of performance issues that SQEX were obviously unimpressed by. So the current official launcher has been stuck at CX18.5 for the past 3-4 years. I suspect that one or more changes I made to MVK in the time since CX18.5 were responsible, but I hadn't been able to fix the performance problems.

The irony here is that the problems with occlusion culling probably were exacerbated by my efforts to *fix* the problem, which is long-standing. Basically, there's a massive impedance mismatch between occlusion culling and D3D11 (and Vulkan) and Metal. This causes problems--I suspect some sort of race condition--where the occlusion counter is somehow wrong. I thought I had fixed them, but clearly the problem is still present.

**Comment by seathasky** [comment]
> > It's very unfortunate situation for Mac FF14 players right now. Square Enix native mac launcher does not work on Monterey nor has it been updated in 2 1/2 years.
> 
> I had a view from inside CW before getting laid off, and... it's probably my fault.
> 
> We tried updating MoltenVK to a newer version, but we encountered all sorts of performance issues that SQEX were obviously unimpressed by. So the current official launcher has been stuck at CX18.5 for the past 3-4 years. I suspect that one or more changes I made to MVK in the time since CX18.5 were responsible, but I hadn't been able to fix the performance problems.
> 
> The irony here is that the problems with occlusion culling probably were exacerbated by my efforts to _fix_ the problem, which is long-standing. Basically, there's a massive impedance mismatch between occlusion culling and D3D11 (and Vulkan) and Metal. This causes problems--I suspect some sort of race condition--where the occlusion counter is somehow wrong. I thought I had fixed them, but clearly the problem is still present.

Thanks for your transparency on the matter! That's very interesting indeed.

**Comment by billhollings** [comment]
> Probably anyone in this thread would be willing to provide more detailed setup assistance

Okay...I've got a SqEx account, downloaded XIV on Mac, and had it download FFXIV.

In order to debug, I want to replace the link to MoltenVK with a link to a debug development version here. Is MoltenVK dynamically linked, and if so, where is the `libMoltenVK.dylib` file? If MoltenVK is statically linked, what would I need to rebuild to link a development MoltenVK, and how would I go about doing that?

**Comment by Gcenx** [comment]
It’s dynamically linked and within a sub directory within the app bundle.

It’s something along the lines of

/Contents/ShareSupport/ (probably ffxiv)/lib64/libMoltenVK.dylib

**Comment by cdavis5e** [comment]
> It’s dynamically linked and within a sub directory within the app bundle.
> 
> It’s something along the lines of
> 
> /Contents/ShareSupport/ (probably ffxiv)/lib64/libMoltenVK.dylib

It's `SharedSupport` with a 'd'.

And I think the third directory name is like `ffxivonline` or some such. It would be `CrossOver` on regular CrossOver, and it's where the port name goes for wrappers like this.

**Comment by zebe** [comment]
Note that modifying the app bundle will break the code signature, so you may need to remove any quarantine bits on it to get it to run, depending on your security settings; you can do this with: `xattr -d com.apple.quarantine [file]`

**Comment by Gcenx** [comment]
@cdavis5e thanks yeah I’m on mobile but figured a quick response would at least get Bill onto the right track.

**Comment by ZoeyCluff** [comment]
@billhollings here's the moltenVK path: /Applications/FINAL FANTASY XIV ONLINE.app/Contents/SharedSupport/finalfantasyxiv/lib64/libMoltenVK.dylib

**Comment by Gcenx** [comment]
Oh we’re all forgetting something important stock MoltenVK won’t work with the shipped DXVK without faking some extensions.

@billhollings you’ll ether need to fake some extensions listed in #203 or swap out the used DXVK dlls with something that works with stock MoltenVK 

If you want to swap out the copy of DXVK grab one of the archives I’m providing from https://github.com/Gcenx/DXVK-macOS/releases and replace the contents of 

/Applications/FINAL FANTASY XIV ONLINE.app/Contents/SharedSupport/finalfantasyxiv/lib64/wine/DXVK

(that’s if that old wrapper symlinked to the dlls)

**Comment by marzent** [comment]
> > Probably anyone in this thread would be willing to provide more detailed setup assistance
> 
> Okay...I've got a SqEx account, downloaded XIV on Mac, and had it download FFXIV.
> 
> In order to debug, I want to replace the link to MoltenVK with a link to a debug development version here. Is MoltenVK dynamically linked, and if so, where is the `libMoltenVK.dylib` file? If MoltenVK is statically linked, what would I need to rebuild to link a development MoltenVK, and how would I go about doing that?

@billhollings The above only applies to the app bundle SE ships which is based on DXVK and Wine components which are almost 3 years old. 

If you are testing with XIV on Mac you can basically ignore everything mentioned, including replacing DXVK. 

MVK is dynamically linked and loaded either from `/Contents/Resources/MoltenVK/[modern|stable]/libMoltenVK.dylib` depending upon the checkbox in the XOM preferences.

**Comment by billhollings** [comment]
> If you are testing with XIV on Mac you can basically ignore everything mentioned, including replacing DXVK.

Thanks. I am able to investigate now using current and patched MoltenVK versions.

Couple of environment questions:

- XIV on Mac is hardened and I can't start it from the Xcode debugger. What's the best way to debug in this environment?
- How can I view the `stderr` output from MoltenVK within the XIV on Mac/FFXIV environment?

**Comment by ZoeyCluff** [comment]
> > If you are testing with XIV on Mac you can basically ignore everything mentioned, including replacing DXVK.
> 
> Thanks. I am able to investigate now using current and patched MoltenVK versions.
> 
> Couple of environment questions:
> 
>     * XIV on Mac is hardened and I can't start it from the Xcode debugger. What's the best way to debug in this environment?

Best bet would probably be to clone the repo and run it through xcode https://github.com/marzent/XIV-on-Mac/ 

>     * How can I view the `stderr` output from MoltenVK within the XIV on Mac/FFXIV environment?

@Gcenx  may know, this one I'm not 100% sure on.

**Comment by Gcenx** [comment]
I haven’t tested but adding `+vulkan` to the winedebug field would provide additional output, this might be more on the wine vulkan side though.

**Comment by billhollings** [comment]
> I believe MoltenVK related messages all get dumped into wine.log

Which is located where?

**Comment by Gcenx** [comment]
> Which is located where?

The logs would be found in `~/Library/Application Support/XIV on Mac/`

**Comment by marzent** [comment]
4.X has more robust logging which should be at `~/Library/Application Support/XIV on Mac/logs/wine.log` (the old NSTask ReadabilityHandler wasn't super reliable at writing wines output to a file)

**Comment by marzent** [comment]
Skipping the loop at https://github.com/KhronosGroup/MoltenVK/blob/a044496a6578bacd7de43aebad547bf0ee4ca40d/MoltenVK/MoltenVK/Commands/MVKCommandEncoderState.mm#L1129 completely fixes the issue but causes some models to not render correctly, while enabling the in-game occlusion culling option.

Running the loop at most once, seems to mitigate this slightly but brings back a tiny amount of flicker.

**Comment by stefand** [comment]
The problem seems to be that the recycled buffers contain old results whereas the re-created buffers are zeroed on creation. Metal isn't clear on either points, but it seems that the visibility counter just increments whatever is already in the result buffer. Also it doesn't say anything about initial contents, not even that they are undefined.

The attached hack fixes the flicker for me. Tested with wined3d, not dxvk though. Please give it a try to confirm my theory. A nicer fix is needed obviously. Making the buffer shared for the purpose of a memset is ugly. I was hoping to pass a load action similar to that of color buffers, but that doesn't seem to be an option...

And yes, it is the memset that fixes this, not a side effect of the changed storage mode.
[zeroinit.txt](https://github.com/KhronosGroup/MoltenVK/files/9686958/zeroinit.txt)

**Comment by reckenrode** [comment]
I applied the patch to my [FFXIV derivation][1], rebuilt it, and ran around a bit and checked the character creator. Unfortunately, I don’t have time to play much right now. However, it looks like the flicking is gone. The character creator in particular is an easy place to test. The clouds in the Las Noscea and Thanalan environments tend to flicker a lot. That’s gone. Even if you change the setting from day to night and back, they still render correctly (or at least without any flickering).

I’m using MoltenVK 1.1.11 (with the above patch), DXVK 1.10.3, and Wine 7.17 all from nixpkgs. I expect it should also work pretty well with XIV on Mac, but I haven’t tested it.

[1]: https://github.com/reckenrode/nixos-configs/blob/main/pkgs/ffxiv/default.nix

**Comment by marzent** [comment]
This patch worked perfectly (tested on master)! 
Given that it was a regression introduced with MVK 1.1.2 it would wonder me greatly if there were any differences in behaviour with different wine versions (even seems to be the same with dxvk vs wined3d)

**Comment by stefand** [comment]
No, I don't expect the Wine version to make a difference. It's simply a case of reusing an old buffer with previous contents vs using a new buffer that is apparently zero initialized.

**Comment by stefand** [comment]
Another experiment with interesting results: Fill the buffer with random data on reuse. If the previous theory that visibility queries just increment old buffer contents is right, this should cause massive flicker. Interestingly it does not. This attached patch fixes the bug just as much as the zeroing bug (while killing performance, due to repeated rand() invocations).

This suggests to me that something else must be going on and touching the buffer avoids the bug due to other reasons - maybe touching the buffer on the CPU side flushes GPU side caches or something like that. Remember, merely making it shared is not enough. The buffer needs to be written to by the GPU (or read? I haven't checked yet).

It looks like a Metal and/or driver bug to me. I've been debugging this on a MacBookPro14,3 with the Radeon GPU. I'll try with my M1 mac and the Intel GPU in the old mbp.

[rand.txt](https://github.com/KhronosGroup/MoltenVK/files/9738673/rand.txt)

**Comment by marzent** [comment]
I believe I have seen users report this problem on most gpu combinations including amd, intel, nvidia and M1 through Rosetta… (Although intel has bigger problems than just the flickering including very grainy dithering everywhere, black screens and random crashes; so much so that we gave up on it (at least with dxvk))

**Comment by stefand** [comment]
It isn't a matter of merely touching the buffer contents. Different operations lead to flicker being fixed or not fixed. I wonder if there are some integer shenanigans going on.

Replacing the memset with this:

```
            uint32_t *contents = (uint32_t *)ba->_mtlBuffer.contents, i;
            for (i = 0; i < ba->_length / sizeof(*contents); i += 1)
                contents[i]++;

```

Surprisingly fixes the flicker. From now on I'll list a few things that fix and don't fix it, but only mention the line inside the for loop:

`                contents[i] &= ~1u;
`
Flicker remains

`
                contents[i] |= 0x80000000u;
`

flicker fixed

`
                contents[i] &= ~0xff000000u;
`

Flicker remains.

Iterating over the buffer in uint64_t's (or just i+=2) with the contents[i]++ loop does not fix the flicker either.

So are there some problems with integer overflows? Are only some bits zeroed by metal when a buffer is reused for another visibility query? I am pretty lost.

**Comment by cdavis5e** [comment]
> I am pretty lost.

IIRC, when we were first investigating this about 3-4 years back, we found out that Metal zeroes out the visibility buffer whenever you start a new render pass. As you know, this is complicated by the fact that MoltenVK often has to interrupt render passes to run the vertex and tessellation control (hull) stages of a tessellated draw, meaning a Vulkan render pass or even a subpass does not have a 1:1 relationship with a Metal render pass. So I had MoltenVK use temporary buffers, and then accumulate the results to their final destination using a compute shader. It appeared to fix the flickering for me at the time, though I did notice that there was some apparently racy behavior that caused the problem to sometimes occur, sometimes not. I thought I fixed those race conditions, but clearly not. If I had to venture a guess, I'd say there's probably still some missing barrier or fence somewhere.

The root cause, of course, is that the query sometimes returns zero (totally occluded) when it should return nonzero.
